### PR TITLE
blog: Predefined models exp

### PR DESCRIPTION
**Pull request by flippedcoder**
This is a post about how you can use DVC experiments to help fine-tune a pre-trained model like AlexNet or SqueezeNet.

### CODE REVIEW

**Comment by jorgeorpinel** [comment]
Cool! Please let us know when you'd like reviewers involved. Thanks

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (jendefig) [inline content/blog/2021-08-24-pre-defined-models-exp.md:57]: insert "we"

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (jendefig) [inline content/blog/2021-08-24-pre-defined-models-exp.md:107]: maybe add: "short for dependencies, defines the ..."  This is probably is clear to the software engineering set, but I think will help clarify for those that are more green.

**Review by dberenbaum — COMMENTED** [review]
Fun article idea!

With the hyperparameter tuning post, it seemed mostly useful to highlight a dvc pattern that wasn't documented elsewhere - how to do hyperparameter searches using dvc experiments. This article keeps to the basic experiments workflow that users can already learn from other docs, so it seems like the value is more about the interest in transfer learning. In that case, interesting results might be more important here, or at least a meaningful approach to how/why to try different experiments. It might make sense to touch base with @daavoo whose background is in computer vision.

Also, could you add attribution to the original source code used to derive this example (and where to get the data)? I see something sort of similar in https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html. One idea for this post would be to update the code to make it more dvc-like and then show how this can make it easier to track different experiments. Some examples:

* Adding `model_name` to the parameters makes it so you can see at a glance whether an experiment is alexnet or squeezenet.
* Logging each epoch using dvclive or using checkpoints enables tracking results per epoch, which currently are getting printed to the screen and then lost.
* Saving both the training and validation metrics shows a more complete picture (it looks like only the training metrics are being saved now) and you can choose to view one or the other or both.

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [inline content/blog/2021-08-24-pre-defined-models-exp.md:5]: 💅 style, make it < 80

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [inline content/blog/2021-08-24-pre-defined-models-exp.md:8]: 2021-08-17 - fix the date?

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [inline content/blog/2021-08-24-pre-defined-models-exp.md:76]: bash -> dvc

**Comment by shcheklein** [comment]
Screen Shot 2021-08-05 at 5 52 59 PM

Something is wrong, does it work for you locally?

https://dvc-org-pre-defined-mod-kedssi.herokuapp.com/blog/pre-defined-models-exp is also broken it looks like ..

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [inline content/blog/2021-08-24-pre-defined-models-exp.md:28]: Should we use this in the title/descriptions? Something "Faster Transfer Learning with Experiments"? (can capture more attention?)

or `Using Experiments for Transfer Learning`

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [inline content/blog/2021-08-24-pre-defined-models-exp.md:23]: good to have links - to transfer learning, AlexNet

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [reply content/blog/2021-08-24-pre-defined-models-exp.md:28]: Smooth Transfer Learning

or 

Pain-free Transfer Learning :)

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [inline content/blog/2021-08-24-pre-defined-models-exp.md:46]: is there some image we can put - with bees and ants :)

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [inline content/blog/2021-08-24-pre-defined-models-exp.md:85]: this not a DVC pipeline yet ... it should be "DVC init"

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [inline content/blog/2021-08-24-pre-defined-models-exp.md:60]: It doesn't look about pipelines, unless I missing something

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [reply content/blog/2021-08-24-pre-defined-models-exp.md:57]: Last two sentences with minor modifications could make it to the `descpritionLong`, which will benefit from it a lot I think.

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [inline content/blog/2021-08-24-pre-defined-models-exp.md:88]: `Put DVC in place` - good name for the section :)

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [inline content/blog/2021-08-24-pre-defined-models-exp.md:101]: so long lines could be split with `\` - please take a look at the docs, there are examples in the pipelines get started

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [inline content/blog/2021-08-24-pre-defined-models-exp.md:118]: We can start with this YAML (and then briefly explain the helper command that we were using to generate it). Also it would be good to explain why do we need this file (for DVC to start managing data, understand where metrics to pay attention to, etc - some friendly explanations)

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (flippedcoder) [reply content/blog/2021-08-24-pre-defined-models-exp.md:50]: Good catch! I got rid of that hyphenated one.

**Comment by flippedcoder** [comment]
> * Adding `model_name` to the parameters makes it so you can see at a glance whether an experiment is alexnet or squeezenet.

Good point! I added that param to the stage creation and in the repo.

**Comment by flippedcoder** [comment]
> * Logging each epoch using dvclive or using checkpoints enables tracking results per epoch, which currently are getting printed to the screen and then lost.

I'll do that! I'm trying to break myself out of thinking in just experiments. 😅

**Comment by flippedcoder** [comment]
> * Saving both the training and validation metrics shows a more complete picture (it looks like only the training metrics are being saved now) and you can choose to view one or the other or both.

Done!

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (flippedcoder) [reply content/blog/2021-08-24-pre-defined-models-exp.md:156]: If you clone the repo, then nope, there's no need for the steps above. I can cut that and just push for them to clone the repo and I'll start doing that for other blog posts.

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (flippedcoder) [reply content/blog/2021-08-24-pre-defined-models-exp.md:100]: 1. Done!
2. Working on adding checkpoints.
3. I'll add the model as an output to the stage.

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (flippedcoder) [reply content/blog/2021-08-24-pre-defined-models-exp.md:28]: I like `Using Experiments for Transfer Learning`!

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (flippedcoder) [reply content/blog/2021-08-24-pre-defined-models-exp.md:46]: That's a good idea! I'll run it by Jeny. :)

**Comment by daavoo** [comment]
> Fun article idea!

I also like the idea very much. 

I would expect the target audience to be someone at least familiar with "finetuning" but that would like to know about how to do it the "DVC way" and my general feedback is based on that assumption, although some things might be also benefitial to other targets. 

Trying to view it from that target perspective, I would love to read something that feels easily reusable for my custom use case. Could be even a way to engage with people at the end of the post with stuff like: now you go and try it with your data, or try to improve our results by switching params, etc.

In that regard, I think that it could be good to dig a little bit in the data preparation phase, using as base the [Data Versioning Tutorial](https://dvc.org/doc/use-cases/versioning-data-and-model-files/tutorial) could be useful for adding references to a more detailed tutorial, where the experimentation part kind of feels outdated. 

In addition, I have no strong preferences regarding *ML Frameworks* but my general feeling is that using a higher level framework (i.e. keras) would reduce the amount of boilerplate code that could make it harder to differenciate the relevant "finetuning with DVC" parts from regular boilerplate.
 
>  It might make sense to touch base with @daavoo whose background is in computer vision.

 It's kind of tricky to know beforehand which experiments would be meaningful 😅  but I would try to share a few ideas/intuitions that might showcase the benefits of experimentation. 

For pretty much every idea, I think the following things would be helpful to support discussions:

- Add per-epoch and per split (train/val) metrics (could be done via DVCLive as mentioned by Dave)
- Add `training time` or `time per epoch` as a metric. This might be tricky for many reasons but even a naive measure would be useful.

Here are those ideas for experimentation "à la DVC":

### Adding more training data

**Depends on**: Having different dataset versions. Having `training_time` as metric.

**Showcases**: DVC data versioning.

**I would expect that**: extending the training set would improve perfomance. Even if that its not the case we might use the results to introduce another useful but less discussed topic: extending the validation set. 

### Comparing model architectures

**Depends on**: Having `model_name` as param. Having `training_time` as metric.

**Showcases**: `DVC params` 

**I would expect that**: 

Using a smaller version (lower `training_time`) of one family of architectures (i.e. `ResNet18`) would have worse validation accuracy than a bigger version of the same family (i.e. `ResNet50`). If that it's not the case might indicate that our problem/dataset is too simple (i.e. need to extend validation dataset to properly showcase this). 

Using a smaller model from newer families (i.e. `MobilenetV3`) would have better or similar accuracy than a bigger model from older families (i.e. `VGG19`)

**Note**: We are currently limiting the models I assume that due to torchvision not being very friendly about finetuning (i.e. can't easily reuse the same "add a new head" code for different architectures) I would either add a few if/elses for different architectures or switch to a framework where the code for adding a new head is more reusable.

### Comparing Finetuning strategies

We could use similar terminology to the one used in https://lightning-flash.readthedocs.io/en/latest/general/finetuning.html#finetune-strategies

**Depends on**: Having `feature_extract` as param. Having `training_time` as metric.

**Showcases**: `DVC params` 

**I would expect that**:  Training the entire model would result in higher `training_time`, for sure.; I'm not sure about `accuracy`. The theory says that, given a large enough dataset, training the entire model would have better perfomance than training only the head.


### Comparing Input Sizes / Data Augmentation

**Depends on**: Adapting the instantiation of `data_transforms` to be dynamic based on input `params` (I can share a code snippet for this). Having access to both `train` and `validation` metrics.

**Showcases**: `DVC params` 

**I would expect that**:  

Using smaller input size would result in lower `training_time` and  potentially lower `accuracy`.

Assuming we remove any data augmentation from the baseline, adding data augmentation (i.e. `RandomHorizontalFlip`) should improve `validation accuracy`.

---

There are probably other things that could be exposed in a similar way to `model_name`, like the Optimizer to be used (`optimizer_name`) however I think that the above ideas would be more understable for users less familiar with Deep Learning  and have more predictable results. 

Smaller vs bigger model, more data vs less data, "fancy new" vs "old school" architecture, etc. feels simpler than `momentum=0.9` vs `momentum=0.5`

**Comment by flippedcoder** [comment]
> Screen Shot 2021-08-05 at 5 52 59 PM
> 
> Something is wrong, does it work for you locally?
> 
> https://dvc-org-pre-defined-mod-kedssi.herokuapp.com/blog/pre-defined-models-exp is also broken it looks like ..

I'm having issues with the `sharp` package locally. Not sure if that's related to this though.

**Comment by shcheklein** [comment]
@flippedcoder 

that's what I see in the build logs:

```
error Error processing Markdown file /tmp/build_6baa4287/content/blog/2021-08-24-pre-defined-models-exp.md:
      can not read a block mapping entry; a multiline key may not be an implicit key at line 6, column 16:
    descriptionLong: |
                   ^
       not finished source and transform nodes - 1.106s
```

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (flippedcoder) [reply content/blog/2021-08-24-pre-defined-models-exp.md:23]: I have some links for AlexNet and SqueezeNet a little later and I added a link explaining transfer learning.

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (flippedcoder) [reply content/blog/2021-08-24-pre-defined-models-exp.md:127]: I have the description of the `dvc stage add` command before this YAML. So you want me to move that description below the YAML? And I added friendlier explanations for each part of the command.

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [inline content/blog/2021-08-24-pre-defined-models-exp.md:7]: 💅 please let's wrap it at 80 symbols please

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [reply content/blog/2021-08-24-pre-defined-models-exp.md:47]: I meant something actual, from the dataset?

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [reply content/blog/2021-08-24-pre-defined-models-exp.md:89]: is it resolved? should we update the git commit message?

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [reply content/blog/2021-08-24-pre-defined-models-exp.md:128]: yep, it can be simler to some extent - YAML is more or less friendly and something folks are already familiar vs some new command they haven't seen before. No strong opinion though. @dberenbaum wdyt? :)

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [inline content/blog/2021-08-24-pre-defined-models-exp.md:86]: git -> dvc

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [reply content/blog/2021-08-24-pre-defined-models-exp.md:23]: yep, it's fine .. usual convention (for docs) is to put link the first time you mention the concept. Not a big deal clearly.

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (shcheklein) [inline content/blog/2021-08-24-pre-defined-models-exp.md:56]: rendering is broken

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (dberenbaum) [reply content/blog/2021-08-24-pre-defined-models-exp.md:128]: Yes, I think the first step could be cloning the repo and then explaining parts of the repo as necessary like `dvc.yaml`. I wouldn't repeat the steps that you've already done in that repo since it could be confusing and it takes longer to get to the transfer learning part. You can also give people the option to skip the explanations and jump right into the transfer learning part. I'm not even sure we really need a detailed explanation of the setup at the start of each blog post, but at least we should give the option of getting right to the main subject matter.

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (flippedcoder) [reply content/blog/2021-08-24-pre-defined-models-exp.md:89]: I removed this section because we all agreed that we don't need to set up DVC in every blog post.

**Comment by flippedcoder** [comment]
@dberenbaum @shcheklein I added checkpoints to the project, updated the explanation of the yaml to match this, and updated the tables. Let me know if there's anything else you see that should be cleaned up or added!

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (dberenbaum) [inline content/blog/2021-08-24-pre-defined-models-exp.md:216]: The downside to adding checkpoints is that it changes the workflow. As you see in the table, `dvc exp run` will continue training additional epochs on top of your previous experiment. To start back at step 0, you will need to do `dvc exp run --reset`.

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (dberenbaum) [inline content/blog/2021-08-24-pre-defined-models-exp.md:202]: Minor suggestion to introduce `-S` here also so that you can shorten all the commands below.

**Review by dberenbaum — CHANGES_REQUESTED** [review]
Looks good, thanks @flippedcoder! I have a few more comments, mostly about either how to keep winnowing down the setup section or else addressing changes introduced by dvclive and checkpoints. IMO it's ready to go once those are fixed.

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (flippedcoder) [reply content/blog/2021-08-24-pre-defined-models-exp.md]: Yeah, I like how that flows better. I got rid of the other link to the original project in favor of this.

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (flippedcoder) [reply content/blog/2021-08-24-pre-defined-models-exp.md:175]: Ope! I pushed my stuff to the repo after running my experiments. 😅 That _should_ be 5 now.

**Comment by jorgeorpinel** [comment]
@dberenbaum @shcheklein @daavoo thanks for reviewing. It looks like @flippedcoder has addressed everything? Hoping for some approvals soon

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (dberenbaum) [inline content/blog/2021-08-24-pre-defined-models-exp.md:196]: Do you think it's worth mentioning `--reset` here or saying that training from scratch will be shown below? Just a thought, but up to you.

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (dberenbaum) [inline content/blog/2021-08-24-pre-defined-models-exp.md:273]: Something happened with this table so that the columns aren't aligning.

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (dberenbaum) [inline content/blog/2021-08-24-pre-defined-models-exp.md:303]: `exp-7aac` in these last two tables looks off to me. It looks like you have squeezenet checkpoints that are running on top of alexnet checkpoints, which doesn't seem like the workflow you have described.

**Comment by dberenbaum** [comment]
There are a couple of table formatting issues I see. Once those look good, I'm happy to approve 🙂 .

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (dberenbaum) [inline content/blog/2021-08-24-pre-defined-models-exp.md:110]: One more thought after taking a quick look at https://github.com/iterative/pretrained-model-demo/blob/132b6928ab11bbdb500e204c18e6491a03944b6b/pretrained_model_tuner.py#L84. The validation metrics are still being manually logged the old way and not captured by dvc. You could consider logging them with dvclive also like:

```python
            if phase == 'train':
                torch.save(model.state_dict(), "model.pt")

                dvclive.log('train_acc', epoch_acc.item())
                dvclive.log('train_loss', epoch_loss)
                dvclive.log('train_time', epoch_time_elapsed)

            if phase == 'val':
                dvclive.log('val_acc', epoch_acc.item())
                dvclive.log('val_loss', epoch_loss)
               
                dvclive.next_step()
```

Unfortunately, that does make your table even wider since you are adding more metrics. Another option might be to only log the validation metrics, since those are usually more helpful than training metrics, or to introduce the `--include-metrics`/`--include-params` flags to narrow the table.

**Comment by dberenbaum** [comment]
Final note: feel free to ping me again to take a final look. I might miss when the comments get resolved otherwise.

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (flippedcoder) [reply content/blog/2021-08-24-pre-defined-models-exp.md:110]: I added those metrics to the table and I took out a few of the params we aren't adjusting.

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (flippedcoder) [reply content/blog/2021-08-24-pre-defined-models-exp.md:303]: Yeah, not sure what I did there but it should be cleared up now!

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (dberenbaum) [reply content/blog/2021-08-24-pre-defined-models-exp.md:110]: Sounds good! Not sure if you are still making updates, but make sure to put `dvclive.next_step()` inside `if phase == 'val'` or else you will be making two steps for every epoch (once for train, once for val).

**Comment by flippedcoder** [comment]
@dberenbaum I fixed all of the tables/workflow stuff!

**Comment by dberenbaum** [comment]
Thanks, @flippedcoder! Did you see my last comment in https://github.com/iterative/dvc.org/pull/2656#discussion_r692150367?

File: content/blog/2021-08-24-pre-defined-models-exp.md
Comment (flippedcoder) [reply content/blog/2021-08-24-pre-defined-models-exp.md]: Oh! That makes sense. I wasn't quite sure about that. I'll make those updates here soon!

Comment (flippedcoder) [reply content/blog/2021-08-24-pre-defined-models-exp.md]: @dberenbaum Just pushed the update for this!

**Comment by dberenbaum** [comment]
Thanks again @flippedcoder! There's still some misalignments between the table headers (starting with the `val_loss` column) and other rows in some of the tables. Are you seeing the same?

**Comment by flippedcoder** [comment]
@dberenbaum I've been fiddling with the table for a while and it seems like it's a combination of the formatting and the number of columns. It lines up a little better without the colors and bolding on rows, otherwise I didn't find anything else I could do about it.

**Comment by dberenbaum** [comment]
@flippedcoder So that's the way it looks when you copy and paste from your terminal? Does the table look fine in your terminal?

IMO it's still worth cleaning up by manually adding/deleting spaces as needed so that it looks like the actual table output. I updated one of the tables in https://github.com/iterative/dvc.org/pull/2656/commits/7ba79bb29dc00a7f42fb8090a66ee429b43a46d5. Could you try to do the same for the tables below?

**Comment by flippedcoder** [comment]
> So that's the way it looks when you copy and paste from your terminal? Does the table look fine in your terminal?

@dberenbaum It's weird because in my terminal it looks fine, but when I apply the formatting code it shifts things. I'll go through and manually update them then!

**Comment by flippedcoder** [comment]
@dberenbaum The problem was with the table headers. It should be all cleared up now!

**Review by dberenbaum — APPROVED** [review]

**Comment by shcheklein** [comment]
There is a a broken link https://github.com/iterative/pretrained-model-demo (project is private - do we want to make it public before merge?)

Could we also try to minify static/uploads/images/2021-08-24/pretrained-models.png ? tinypng or some other similar service works.

**Comment by jendefig** [comment]
> Could we also try to minify static/uploads/images/2021-08-24/pretrained-models.png ? tinypng or some other similar service works.

@shcheklein I just ran the image through tinypng a second time.  It shaved off another 2%.  I guess it's bigger than normal because it has all the images in it?  Roman had already compressed it.

**Comment by flippedcoder** [comment]
> There is a a broken link https://github.com/iterative/pretrained-model-demo (project is private - do we want to make it public before merge?)

@shcheklein I just changed the repo to public and it looks like everything's good to go! Would you mind merging this?

**Comment by shcheklein** [comment]
@flippedcoder done, merged 🎉 !

**Pull request merged**
